File: src/Will_cpp_pkg/src/turtle_controller.cpp
#include "rclcpp/rclcpp.hpp"
#include "my_robot_interfaces/msg/turtle2_array.hpp"
#include "my_robot_interfaces/srv/catch_turtle.hpp"
#include "turtlesim/msg/pose.hpp"
#include "geometry_msgs/msg/twist.hpp"
#include <random>
#include <cmath>

using namespace std::placeholders;


class TurtleControllerNode : public rclcpp::Node
{
public:
    TurtleControllerNode() : Node("turtle_controller") 
    {
        this->declare_parameter("catch_closest_turtle", true);
        catch_closest_turtle_first = this->get_parameter("catch_closest_turtle").as_bool();
        subscriber_pose_ = this->create_subscription<turtlesim::msg::Pose>("turtle1/pose", 10, std::bind(&TurtleControllerNode::callbackPose, this, _1));
        subscriber_alive_turtle_ = this->create_subscription<my_robot_interfaces::msg::Turtle2Array>(
            "alive_turtle", 10, std::bind(&TurtleControllerNode::turtleList, this, _1));
        publisher_ = this->create_publisher<geometry_msgs::msg::Twist>("turtle1/cmd_vel", 10);
        timer_ = this->create_wall_timer(std::chrono::milliseconds(100), std::bind(&TurtleControllerNode::chaseTurtle, this));
        catch_turtle_ = this->create_client<my_robot_interfaces::srv::CatchTurtle>("catch_turtle", 10);
    }
private:
    void callbackPose(const turtlesim::msg::Pose::SharedPtr msg)
    {
        turtle1_x = msg->x;
        turtle1_y = msg->y;
        turtle1_theta = msg->theta;
    }

    void turtleList(const my_robot_interfaces::msg::Turtle2Array::SharedPtr msg)
    {   
        alive_turtles_ = msg->turtles;
    }

    void chaseTurtle()
    {
        if (alive_turtles_.empty())
            return;
        if (catch_closest_turtle_first)
        {
            if (!target_exist) 
            {
                double min_dis = 999;
                int index = 0;
                for (size_t i = 0; i < alive_turtles_.size(); i++)
                {
                    distance = sqrt(pow((alive_turtles_[i].x-turtle1_x), 2.0) + pow((alive_turtles_[i].y-turtle1_y), 2.0));
                    if (distance < min_dis)
                    {    
                        min_dis = distance;
                        index = i;
                    }
                }
                target_x = alive_turtles_[index].x;
                target_y = alive_turtles_[index].y;
                target_name = alive_turtles_[index].name;
                RCLCPP_INFO(this->get_logger(),"%s !", target_name.c_str());
                target_exist = true; 
            }
        }
        else
        {
            if (!target_exist) 
            {
                std::random_device rd;
                std::mt19937 gen(rd());
                std::uniform_int_distribution<int> dist(0, alive_turtles_.size()-1);
                int index = dist(gen);
                target_x = alive_turtles_[index].x;
                target_y = alive_turtles_[index].y;
                target_name = alive_turtles_[index].name;
                RCLCPP_INFO(this->get_logger(),"%s !", target_name.c_str());
                target_exist = true; 
            }
        }
        
        distance = sqrt(pow((target_x-turtle1_x), 2.0) + pow((target_y-turtle1_y), 2.0));
        dis_theta = atan2((target_y-turtle1_y), (target_x-turtle1_x))-turtle1_theta;
        if (dis_theta > M_PI)
            dis_theta -= 2 * M_PI;
        else if (dis_theta < -M_PI)
            dis_theta += 2 * M_PI;
        vx = 2.5 * distance;
        az = 4.5 * dis_theta;
        auto cmd_vel = geometry_msgs::msg::Twist();  
        cmd_vel.linear.x = vx;
        cmd_vel.angular.z = az;
        if (distance > 0.5)
            publisher_->publish(cmd_vel);
        else
        {
            RCLCPP_INFO(this->get_logger(),"Reach target !");
            cmd_vel.linear.x = 0;
            cmd_vel.angular.z = 0; 
            publisher_->publish(cmd_vel);
            auto request = std::make_shared<my_robot_interfaces::srv::CatchTurtle::Request>();
            request->name = target_name;
            catch_turtle_->async_send_request(
                        request,
                        [this, request](rclcpp::Client<my_robot_interfaces::srv::CatchTurtle>::SharedFuture future)
                    {
                        auto response = future.get();
                        if (response)
                            {
                                RCLCPP_INFO(this->get_logger(),"%s has been catched !", request->name.c_str());
                                target_exist= false;    
                            }    
                        
                    });

                  
        }
    }

    rclcpp::Subscription<turtlesim::msg::Pose>::SharedPtr subscriber_pose_;
    rclcpp::Subscription<my_robot_interfaces::msg::Turtle2Array>::SharedPtr subscriber_alive_turtle_;
    rclcpp::Publisher<geometry_msgs::msg::Twist>::SharedPtr publisher_;
    rclcpp::TimerBase::SharedPtr timer_;
    std::vector<my_robot_interfaces::msg::Turtle2> alive_turtles_;
    rclcpp::Client<my_robot_interfaces::srv::CatchTurtle>::SharedPtr catch_turtle_;
    bool target_exist= false;
    float turtle1_x, turtle1_y, turtle1_theta;
    double target_x, target_y, distance, dis_theta, vx, az;
    std::string target_name;
    bool catch_closest_turtle_first;
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<TurtleControllerNode>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: src/turtlesim_catch_them_all/turtlesim_catch_them_all/turtle_controller.py
#!/usr/bin/env python3 
import rclpy 
from rclpy.node import Node 
import random
import math

from geometry_msgs.msg import Twist
from turtlesim.msg import Pose
from my_robot_interfaces.msg import Turtle
from my_robot_interfaces.srv import CatchTurtle

class TurtleControllerNode(Node):
    def __init__(self):
        super().__init__("turtle_controller")
        self.declare_parameter("catch_closest_turtle_first", False)
        self.catch_closest_turtle_ = self.get_parameter("catch_closest_turtle_first").value
        self.turtles_ = None
        self.target_ = None
        self.turtle_name_ = None
        self.Kp_ = 3.0
        self.waiting_for_response_ = False
        self.pose_subscrber_ = self.create_subscription(Pose, "turtle1/pose", self.callback_pose_subscriber, 10)
        self.alive_turtles_subscriber_ = self.create_subscription(Turtle, "alive_turtles", self.callback_alive_turtles, 10)
        self.cmd_vel_publisher_ = self.create_publisher(Twist, "turtle1/cmd_vel", 10)
        self.move_timer_ = self.create_timer(0.05, self.publish_cmd_vel)
        self.catch_turtle_client_ = self.create_client(CatchTurtle, "catch_turtle")

    def callback_pose_subscriber(self, msg: Pose):
        self.turtle1_X_ = msg.x
        self.turtle1_Y_ = msg.y
        self.turtle1_deg_ = msg.theta
        
    def callback_alive_turtles(self, msg:Turtle):
        self.turtles_ = msg.name
        self.turtles_x_ = msg.x
        self.turtles_y_ = msg.y
        if not self.turtle_name_:
            return
        if self.turtle_name_ not in self.turtles_:
            self.get_logger().info(self.turtle_name_ + " has been killed")
            self.target_ = None
            self.waiting_for_response_ = False

    def publish_cmd_vel(self):
        if self.waiting_for_response_:
            return        
        if not self.turtles_ :
            return
        if not self.target_ :
            if not self.catch_closest_turtle_:
                self.target_ = random.randint(1,len(self.turtles_))
                self.turtle_name_ = self.turtles_[self.target_-1]
            else:
                distance_closest = None
                for name, x, y in zip(self.turtles_, self.turtles_x_,self.turtles_y_):
                    dis_x = round(x - self.turtle1_X_, 1)
                    dis_y = round(y - self.turtle1_Y_, 1)
                    distance = math.sqrt(dis_x**2 + dis_y**2)
                    if distance_closest == None or distance < distance_closest:
                        distance_closest = distance
                        self.turtle_name_ = name
                self.get_logger().info("Target : " + str(self.turtle_name_))
        position = self.turtles_.index(self.turtle_name_)
        target_X = self.turtles_x_[position]
        target_Y = self.turtles_y_[position]
        delta_distance, delta_degree = self.VelocityCalculator(target_X, target_Y)        
        msg = Twist()
        if abs(delta_degree) < 0.5:
            msg.linear.x = self.Kp_ * delta_distance
            msg.angular.z = self. Kp_ * delta_degree
        else:
            msg.linear.x = 0.5 * self.Kp_ * delta_distance
            msg.angular.z = self. Kp_ * delta_degree
        self.cmd_vel_publisher_.publish(msg)

        if delta_distance < 0.5:
            request = CatchTurtle.Request()
            request.name = self.turtle_name_
            self.waiting_for_response_ = True
            future = self.catch_turtle_client_.call_async(request)
            future.add_done_callback(self.callback_catch_turtle)

    def callback_catch_turtle(self, future):
        response = future.result()
        if response:
            pass

    def VelocityCalculator(self, x, y):
        dis_x = round(x - self.turtle1_X_, 1)
        dis_y = round(y - self.turtle1_Y_, 1)
        delta_distance = math.sqrt(dis_x**2 + dis_y**2)
        delta_degree = math.atan2(dis_y, dis_x) - self.turtle1_deg_
        if delta_degree > math.pi:
            delta_degree -= 2 * math.pi
        elif delta_degree < -math.pi:
            delta_degree += 2 * math.pi
        return delta_distance, delta_degree



def main(args=None): 
    rclpy.init(args=args) 
    node = TurtleControllerNode()  
    rclpy.spin(node)    
    rclpy.shutdown() 

if __name__ == "__main__":  
    main()